### DISCUSSION TITLE
feature request: api route global error handling

### DISCUSSION THREAD

**Discussion by y-nk** [vercel/next.js#49984] [Ideas]
### Goals

1. normalize how errors are handled in api routes

### Non-Goals

1. create a complex api that answers all the usecases

### Background

It'd be nice to have an integrated way to define all errors based on code into the framework. Somehow it does it already, since any error thrown is handled as a 500, but we can't customize this error ourselves. We need to return a response with a status code, which isn't always convenient.

## Least effort: extend `NextResponse.error`

At least, Next should extend the `NextResponse.error` to support status code and messages so we could do: `return NextResponse.error(401, 'Unauthorized')`. The bad thing about this is that there's no "bubbling" of the error, so it's not very convenient.

## Best effort: catch `ApiError` instances at framework level

Allowing a user to write `throw new ApiError(401, 'Unauthorized')` which would be handled by Next at a global level. It seems a low hanging fruit with a lot of benefits in DX. We may end up wrapping lib calls in try catch to convert the errors to ApiError instances, but I think it's ok.

## Handle-all-cases: provide an `errorMap` function to Next

Another possible (advanced) solution would be to define a `errorMap` function such as `function errorMap(error: unknown): { statusCode: number, message?: string }`. It's similar to zod's error map, which is global as well. It's super powerful and probably best, but a bit overkill since it may blow up in size.

For example, we could be doing:

```ts
import { setApiErrorMap } from 'next/server'

setApiErrorMap((err) => {
  if (err instanceof ZodError)
    return { statusCode: 400, message: 'bad request' }

  if (err.message === 'unauthorized')
    return { statusCode: 401, message: 'unauthorized' }

  return { statusCode: 500 }
})
```

### I personally prefer the "best effort" solution which seems to be the sweet spot between "not convenient" and "too powerful"

---

In the meantime:

```ts
import { NextResponse } from "next/server";
import { ApiError } from "next/dist/server/api-utils";

type ApiRoute = (req: Request, params: Record<string, string>) => Response | Promise<Response>

export function catchAll(route: ApiRoute): ApiRoute {
  return async function catchAllHandler(req, params) {
    try {
      return route(req, params)
    } catch (err) {
      let statusCode = 500
      let message = 'server error'

      if (err instanceof ApiError) {
        statusCode = err.statusCode
        message = err.message
      } else if (err instanceof Error) {
        message = err.message
      }

      return new Response(message, { status: statusCode })
    }
  }
}
```

**Comment by coder-xiaotian** [comment]
I need this feature!
